src/routes/teams.rs

```
@@ -1,25 +1,275 @@
+        Err(ApiError::AuthenticationError)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct EditTeamMember {
+    pub permissions: Option<Permissions>,
+    pub role: Option<String>,
+    pub name: Option<String>,
+}
+
+#[patch("{id}/members/{user_id}")]
+pub async fn edit_team_member(
+    req: HttpRequest,
+    info: web::Path<(TeamId, UserId)>,
+    pool: web::Data<PgPool>,
+    edit_member: web::Json<EditTeamMember>,
+) -> Result<HttpResponse, ApiError> {
+    let ids = info.into_inner();
+    let id = ids.0.into();
+    let user_id = ids.1.into();
+
+    let current_user = get_user_from_headers(req.headers(), &**pool)
+        .await
+        .map_err(|_| ApiError::AuthenticationError)?;
+    let team_member = TeamMember::get_from_user_id(id, current_user.id.into(), &**pool).await?;
+
+    let mut transaction = pool
+        .begin()
+        .await
+        .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+    if let Some(member) = team_member {
+        if member.permissions.contains(Permissions::EDIT_MEMBER)
+            && member.accepted
+            && edit_member.role.as_deref() != Some(crate::models::teams::OWNER_ROLE)
+        {
+            if let Some(new_permissions) = edit_member.permissions {
+                if !member.permissions.contains(new_permissions) {
+                    return Err(ApiError::AuthenticationError);
+                }
+            }
+
+            TeamMember::edit_team_member(
+                id,
+                user_id,
+                edit_member.permissions,
+                edit_member.role.clone(),
+                None,
+                edit_member.name.clone(),
+                &mut transaction,
+            )
+            .await?;
+
+            transaction
+                .commit()
+                .await
+                .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+            Ok(HttpResponse::Ok().body(""))
+        } else {
+            Err(ApiError::AuthenticationError)
```

See above, this should return a useful error message.